pkg/sidecar/appclone.go

```diff
@@ -36,32 +64,65 @@ func RunCloneCommand(cfg *Config) error {
 		return fmt.Errorf("removing lost+found: %s", err)
 	}
 
-	if cfg.ServerID() > cfg.MyServerIDOffset {
-		// cloning from prior node
-		sourceHost := cfg.FQDNForServer(cfg.ServerID() - 1)
-		err := cloneFromSource(cfg, sourceHost)
-		if err != nil {
-			return fmt.Errorf("failed to clone from %s, err: %s", sourceHost, err)
+	if isServiceAvailable(cfg.ReplicasFQDN()) {
+		if err := attemptClone(cfg, cfg.ReplicasFQDN()); err != nil {
+			return fmt.Errorf("cloning from healthy replicas failed due to unexpected error: %s", err)
+		}
+	} else if isServiceAvailable(cfg.MasterFQDN()) {
+		log.Info("healthy replica service was unavailable for cloning, will attempt to clone from the master")
+		if err := attemptClone(cfg, cfg.MasterFQDN()); err != nil {
+			return fmt.Errorf("cloning from master service failed due to unexpected error: %s", err)
 		}
 	} else if cfg.ShouldCloneFromBucket() {
 		// cloning from provided initBucketURL
-		err := cloneFromBucket(cfg.InitBucketURL)
-		if err != nil {
+		log.Info("cloning from bucket")
+		if err := cloneFromBucket(cfg.InitBucketURL); err != nil {
 			return fmt.Errorf("failed to clone from bucket, err: %s", err)
 		}
 	} else {
-		log.Info("nothing to clone or init from")
+		log.Info("nothing to clone from: no existing data found, no replicas and no master available, and no clone bucket url found")
 		return nil
 	}
 
 	// prepare backup
-	if err := xtrabackupPreperData(); err != nil {
+	if err := xtrabackupPrepareData(); err != nil {
 		return err
 	}
 
 	return nil
 }
 
+func isServiceAvailable(svc string) bool {
```

Comment:
Originally, we tried to just hit the backup endpoint directly, without checking `serverProbeEndpoint` first. However, it was difficult to discern the difference between an empty service and a service that threw an error in the middle of a streaming backup. We of course want to handle those two cases differently.

Without the availability check, determining if the service is empty means inspecting the error for `connection refused` or a specific timeout error, which is error prone itself. With this availability check, we can quickly and cleanly move to the next service if this one is not responding after the 5s timeout.